core/services/relay/evm/codec/codec.go

```diff
@@ -159,3 +164,100 @@ func decodeAddress(data any) (any, error) {
 
 	return common.Address(decoded), nil
 }
+
+// addressStringDecodeHook is a decode hook that converts between `from` and `to` types involving string and common.Address types.
+// It handles the following conversions:
+// 1. `from` string or *string -> `to` common.Address or *common.Address
+// 2. `from` common.Address or *common.Address -> `to` string or *string
+//
+// The function handles invalid `from` values and `nil` pointers:
+//   - If `from` is a string or *string and is invalid (e.g., an empty string or a non-hex string),
+//     it returns an appropriate error (types.ErrInvalidType).
+//   - If `from` is an empty common.Address{} or *common.Address, the function returns an error
+//     (types.ErrInvalidType) instead of treating it as the zero address ("0x0000000000000000000000000000000000000000").
+//   - If `from` is a nil *string or nil *common.Address, the function returns an error (types.ErrInvalidType).
+//
+// For unsupported `from` and `to` conversions, the function returns the original value unchanged.
+func addressStringDecodeHook(from reflect.Type, to reflect.Type, value any) (any, error) {
+	// Helper variables for types to improve readability
+	stringType, stringPtrType := reflect.TypeOf(""), reflect.PointerTo(reflect.TypeOf(""))
+	addressType, addressPtrType := reflect.TypeOf(common.Address{}), reflect.TypeOf(&common.Address{})
```

Comment:
These can be constants outside the function scope. This would save them being declared every time the function runs.